### PR TITLE
WIP: Also compile Binaryen to WebAssembly

### PR DESCRIPTION
**Pull request by dcodeIO**
This PR adds another build target to `build-js.sh` building a WebAssembly version using the JS wrapper as `binaryen-wasm.js`.

It already works (can run the kitchen-sink.js test) under the following conditions:

* When using node.js, `binaryen-wasm.js` must be required from the `./bin` directory because it otherwise can't find `binaryen-wasm.wasm` next to it (looks in the current working directory instead?).
* <del>When not using node.js (which seems to synchronously instantiate the module working around this issue), there is no mechanism yet to tell when the module is ready.</del>

Opening this PR so we can figure something out :)

See also: https://github.com/WebAssembly/binaryen/issues/1380

### CODE REVIEW

**Comment by dcodeIO** [comment]
One way to make loading lazily work could be to provide the `.then` interface [under all conditions](https://github.com/kripken/emscripten/blob/fcab695dc6dea96a1b0a8c26b4e6243a1d8deb2a/src/postamble.js#L121-L146), making it resolve instantly when it's not a lazy build. A non-lazy build would then initialize instantly while a lazy-build would require that the user uses `.then`. Both versions work like `MODULARIZE_INSTANCE`, and the safest way to use it would be to always use `.then` (though not necesary in a non-lazy-build), so things are interchangable.

**Comment by dcodeIO** [comment]
Ok, this one should "just work", but `.then` is named `.ready` here so it doesn't conflict for now.

```js
var Binaryen = ...;

// can test if already ready
console.log(Binaryen.isReady);

// or just use a promise and wait for it
Binaryen.ready.then(function() {
  ...
}, function(err) { throw err; });
```

**Comment by dcodeIO** [comment]
Actually I figured that `.ready` should probably remain a binaryen.js-only thing because it requires specific knowledge of which parts in post-js must be delayed (here: `initializeConstants`). Generalizing this functionality would most likely result in more confusion than it'd be useful. At least "you should use MODULARIZE instead" seems like a more useful advise to me than "you first have to find out which parts of your code would fail in an asynchronous instantiation scenario, then wrap that in a function and do X, then Y and maybe Z".

Remains the cwd issue.

**Comment by kripken** [comment]
For the cwd, you can use `Module.locateFile` to customize where to look. I think that could work, but it would need to check if the file is in the current dir or under `bin/` manually I suppose.

**Comment by kripken** [comment]
Maybe we could use `SINGLE_FILE` to embed the wasm? Not great for size, though.

How about disabling async compilation, does that work in node? (I know it would fail in chrome, but maybe v8 in node.js is more tolerant of sync compilation?)

**Comment by dcodeIO** [comment]
Hmm, I really thought `Module.locateFile` or `Module.filePackagePrefixURL` could solve this (if the latter also works for non-prepackaged files, just trying it as well), but it seems that both aren't actually accessed when I put the following in `binaryen.js-pre.js` and include it in just the wasm build:

```js
Module['locateFile'] = function(file) {
  throw "locateFile called: " + file;
};
Object.defineProperty(Module, 'filePackagePrefixURL', {
  get: function() {
    throw "filePackagePrefixURL accessed";
  }
});
```

Any ideas?

**Comment by kripken** [comment]
The place where it should call `locateFile` is here: https://github.com/kripken/emscripten/blob/incoming/src/preamble.js#L1990 maybe some debugging logging around that location can help.

**Comment by dcodeIO** [comment]
Seems `integrateWasmJS` is executed before the pre-js code is. Is there a way to insert code even earlier?

**Comment by kripken** [comment]
That.. would be very odd if true :) the pre-js goes before the "prelude" file which is where the integrate method is.

Is the pre-js queuing code to happen asynchronously later perhaps?

**Comment by dcodeIO** [comment]
From the unoptimized binaryen-wasm.js I learned:

```
// --pre-jses are emitted after the Module integration code, so that they can
// refer to Module (if they choose; they can also define Module)
```

My pre-js is pretty much exactly

```js
Module['locateFile'] = function(file) {
  throw "locateFile called: " + file;
};
Object.defineProperty(Module, 'filePackagePrefixURL', {
  get: function() {
    throw "filePackagePrefixURL accessed";
  }
});
```

And it is inserted into the final file (line 7537 right after `// {{MODULE_ADDITIONS}}` in top-level execusion) below the call to `integrateWasmJS();` (line 1815 right below `// === Body ===` in top-level execution). The logging I put into both also showed that it's actually executed in that order. Did I find a bug? :)

File: build-js.sh [hunk 1]
```diff
@@ -587,26 +587,29 @@ echo "building binaryen.js"
...
+  "-I$BINARYEN_SRC/" \
   -s EXPORTED_FUNCTIONS=[${EXPORTED_FUNCTIONS}] \
-  -o $BINARYEN_BIN/binaryen-wasm${OUT_FILE_SUFFIX}.js \
+  -o "$BINARYEN_BIN/binaryen-wasm${OUT_FILE_SUFFIX}.js" \
   -s BINARYEN=1 \
   -s 'BINARYEN_METHOD="native-wasm"' \
   -s MODULARIZE_INSTANCE=1 \
   -s 'EXPORT_NAME="Binaryen"' \
-  --post-js $BINARYEN_SRC/js/binaryen.js-post.js
+  -s NO_EXIT_RUNTIME=0 \
```
Comment (dcodeIO) [inline build-js.sh]: Without, there is a warning when running binaryen-wasm.js. Is that correct?

Comment (kripken) [reply build-js.sh]: Something about `atexits`? That's expected when assertions are on, but not in an optimized build. So we shouldn't see it here, I don't think. Or was it something else?

File: src/js/binaryen.js-pre.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Make sure node is able to find the .wasm next to the .js file. Only relevant
+// when building binaryen-wasm.js (not used for binaryen.js currently).
+if (
```
Comment (kripken) [inline src/js/binaryen.js-pre.js]: as this is a pre-js, you can actually just check the `ENVIRONMENT_IS_NODE` variable

Comment (dcodeIO) [reply src/js/binaryen.js-pre.js]: I tried, but it wasn't defined already at this point.

File: build-js.sh (hunk 1 above)
Comment (dcodeIO) [reply build-js.sh]: Let me double check

Comment (dcodeIO) [reply build-js.sh]: You are right, this must have been caused by something else.

File: src/js/binaryen.js-pre.js (hunk 2 above)
Comment (dcodeIO) [reply src/js/binaryen.js-pre.js]: Just checked again without a minifier. Appears pre-js goes right after `if (!Module) Module = typeof Binaryen !== 'undefined' ? Binaryen : {};` (starts at line 21) while `ENVIRONMENT_IS_NODE` is defined a bit later (line 63).

Comment (kripken) [reply src/js/binaryen.js-pre.js]: Oh right, the pre-js goes before everything...

File: check.py [hunk 3]
```diff
@@ -357,28 +357,34 @@ def run_binaryen_js_tests():
...
-      f.write(node_test_glue())
     test_path = os.path.join(options.binaryen_test, 'binaryen.js', s)
     test_src = open(test_path).read()
-    f.write(test_src)
-    f.close()
-    def test(engine):
-      cmd = [engine, 'a.js']
+    def test(engine, runner):
+      print runner + ": " + s
+      f = open(os.path.join(options.binaryen_bin, 'a.js'), 'w')
```
Comment (kripken) [inline check.py]: adding files to `bin/` should be avoided. one option is to clean up those files, but then there's a risk of not reaching the cleanup code. how about copying the js and wasm files into the current dir? if we forget them there, that's less bad than leaving them in the bin dir. Even better would be to do this in a new subdir just for testing, perhaps, like `test/temp/`.

**Comment by kripken** [comment]
About the synchronous/asynchronous issues,

1. Did we check if node.js can handle synchronous? that would be best.
2. Is there some way in module loaders to handle this? What i mean is, when you do require() in node.js, is that an async or a sync operation from the perspective of the module being loaded? If node allowed an async callback for the require to complete, that would be really cool...
3. If we can't do either of those things to keep the API synchronous, I think we should consider changing it to be asynchronous everywhere - the asm.js and wasm versions being different seems confusing. But i'm not sure.

File: check.py (hunk 3 above)
Comment (dcodeIO) [reply check.py]: The current dir is the project's base directory, where `a.js` was generated before, which wasn't that great either. Hence, a temp directory sounds ideal, will take a look.

**Comment by dcodeIO** [comment]
> Did we check if node.js can handle synchronous compilation? that would be best.

I know that node.js supports `new WebAssembly.Instance(new WebAssembly.Module(...), {})`, if that counts as synchronous. Though, as node.js is all about doing things asynchronous, I'd say that designing things to be do such things in the background is even better.

> Is there some way in module loaders to handle this? What i mean is, when you do require() in node.js, is that an async or a sync operation from the perspective of the module being loaded? If node allowed an async callback for the require to complete, that would be really cool...

`require` is a synchronous operation that usually returns something. In `binaryen-wasm`'s case, it returns the same thing a browser would get when accessing the `Binaryen` before it is ready. In all cases, the recommended usage is:

```js
Binaryen.ready.then(() => {
  ... use it ...
});
```

> If we can't do either of those things to keep the API synchronous, I think we should consider changing it to be asynchronous everywhere - the asm.js and wasm versions being different seems confusing. But i'm not sure.

If I understood correctly, there is no way to do asynchronous loading in browsers anyway if one wants to use an additional .wasm file (not embedded as base64 or something), is this right? If so, the default case should be asynchronous anyway because there's at least one important target that doesn't support synchronous instantiation.

From `binaryen-wasm`' vs. `binaryen.js` perspective, the key difference is that `binaryen.js` can, technically, be used right away (conveniently keeps backwards compatibility). BUT: In the docs I'd always recommend the asynchronous way (see example above) from this point onwards, to minimize confusion.

Or, in other words: Switching synchronous `binaryen.js` with asynchronous `binaryen-wasm.js` and vice-versa will just work as long as `Binaryen.ready` is used as shown above. Again, I hope this makes sense :)

File: src/js/binaryen.js-pre.js (hunk 2 above)
Comment (dcodeIO) [reply src/js/binaryen.js-pre.js]: Would it make sense to move just the `ENVIRONMENT_IS_XY` definitions up so pre-js can use/modify them? Could create a PR for this if it makes sense.

Comment (dcodeIO) [reply src/js/binaryen.js-pre.js]: See: https://github.com/kripken/emscripten/pull/6137

**Comment by kripken** [comment]
Yeah, makes sense overall.

About `.ready.then()`, could it be just `.then()` like MODULARIZE does now? Or is `.ready` better for some reason?

I see that Parcel allows [synchronous syntax for loading wasm](https://medium.com/@devongovett/parcel-v1-5-0-released-source-maps-webassembly-rust-and-more-3a6385e43b95), by translating it to async at bundling time. We probably don't want to depend on a specific bundler, though, but that would have been nice...

**Comment by dcodeIO** [comment]
Currently it looks like this:

```js
const { Binaryen } = import "./binaryen-wasm.js";
Binaryen.ready.then(...);
```

or

```js
const { Binaryen } = import "./binaryen-wasm.js";
await Binaryen.ready;
...
```

> About .ready.then(), could it be just .then() like MODULARIZE does now? Or is .ready better for some reason?

`.ready` has the advantage that it is a promise and thus can be used with `await`. It also does some binaryen.js specific things as outlined above, which conveniently retains backwards-compatibility when continuing to use just `binaryen.js`. It might be possible to make `.then` work similar, but that'd require that an Emscripten module instance extends promise and that it also does that in MODULARIZE_INSTANCE. Would look like:

```js
const { Binaryen } = import "./binaryen-wasm.js";
Binaryen.then(...);
```

or

```js
const { Binaryen } = import "./binaryen-wasm.js";
await Binaryen;
...
```

> I see that Parcel allows synchronous syntax for loading wasm, by translating it to async at bundling time. We probably don't want to depend on a specific bundler, though, but that would have been nice...

I don't think that we can hook to `import` like in Parcel's examples, because when using binaryen.js or binaryen-wasm.js one is actually importing a .js file, not a .wasm. Unfortunately, I don't know what happens when the imported element in an `async import(...)` scenario is a promise itself. If it propagates that to where `async import(...)` is happening, changing `.then` as above might actually make it possible, but I'd not bet that this is how it's done.

**Comment by dcodeIO** [comment]
Looked this up pretty quick: It seems we can actually do that import thing. The only thing necessary is to return a promise as the module's export. Does this already work with just `.then` because it resembles one?

**Comment by kripken** [comment]
I think the Modularize `.then()` just looks like a promise, but isn't actually one. This caused some issues in https://github.com/kripken/emscripten/issues/5820 so maybe as discussed there we should change it.

Anyhow, I don't want to block this PR on all these larger issues. But picking the API (with `.ready` or not) is maybe something we should figure out before landing. Or at least we should have an idea of a way to change things later without breakage.

**Comment by dcodeIO** [comment]
I've worked something out [here](https://gist.github.com/dcodeIO/ea3e79fa918a17e027c23d4393c6cf4c), which essentially makes every module a valid promise by means of just a prototypical extension. Let me know if that's something you'd *tolerate* in Emscripten or not, as it's certainly doing... well... JavaScript-y things.

**Comment by dcodeIO** [comment]
The obvious alternative here is to simply provide a common `.ready` property on everything.

* MODULARIZE .ready promises the instantiated module (what the outer wrapper function did) and should then propagate to...
* MODULARIZE_INSTANCE .ready promises the module instance itself once it's ready (what onRuntimeInitialized did).

In any case, the result of .ready would be a ready to use module. In MODULARIZE it is a unique one, with its own memory etc., while in MODULARIZE_INSTANCE it's always the same one.

**Comment by dcodeIO** [comment]
I think I'd favor the `.ready` promise because of its simplicity. Less code and no need for promise shenanigans. Has zero drawbacks except style.

**Comment by kripken** [comment]
Interesting about wabt using `.ready` too - is that a general convention in the JS space these days? If so, definitely let's do that.

**Comment by dcodeIO** [comment]
> is that a general convention in the JS space these days?

Not that I'm aware of. I guess it just makes sense to use a Promise, and to use `ready` as its name :)

**Comment by kripken** [comment]
I see. Well, I don't really have strong thoughts on the API one way or another, just trying to keep us consistent with other stuff... :) I'm ok to merge this in unless you think there's more to discuss.

I am somewhat worried about us already having a non-`ready` method on MODULARIZE, so maybe it's worth opening an issue on emscripten to discuss what to do more generally there.

**Comment by dcodeIO** [comment]
Superseded by https://github.com/WebAssembly/binaryen/pull/2503

**Pull request closed without merging**